### PR TITLE
feat: add Mistral AI provider client (Phase 4.2 Provider Expansion)

### LINKED ISSUE #660: [ENHANCEMENT] Add Mistral AI provider support (Phase 4.2 — Provider Expansion)
### Current Behavior

llm4s currently supports 9 LLM providers: OpenAI, Azure, Anthropic, OpenRouter, Ollama, Zai, Gemini, DeepSeek, and Cohere. As outlined in the [Agent Framework Roadmap (Phase 4.2)](https://github.com/llm4s/llm4s/blob/main/docs/design/agent-framework-roadmap.md), Mistral AI is listed as a planned provider expansion but is not yet implemented.

Users who want to use Mistral models (Mistral Large, Mistral Small, Codestral, etc.) currently have no direct way to do so through llm4s.

### Proposed Improvement

Add a MistralClient provider following the established provider pattern with:

MistralConfig case class (apiKey, model, baseUrl, contextWindow, reserveCompletion) with model-specific context window fallbacks (128K for Large/Small, 256K for Codestral, 32K for Medium/Tiny)
MistralClient extending LLMClient with MetricsRecording, using Mistral's OpenAI-compatible /v1/chat/completions endpoint
case object Mistral in LLMProvider sealed trait with name, fromName, and all list registration
Registration in LLMConnect.buildClient and LLMConnect.getClient pattern matches
Streaming marked as unsupported in v1 (returns ConfigurationError), consistent with the Cohere provider approach

### Benefits & Use Cases

- **Direct Mistral access** — No need for OpenRouter proxy; lower latency and simpler auth
- **Model variety** — Access to Mistral Large (128K context), Mistral Small, Codestral (256K context for code), and open-source Mistral models
- **European AI provider** — Useful for GDPR-compliant deployments since Mistral is EU-based
- **Provider parity** — Fulfills Phase 4.2 (Provider Expansion) of the [agent-framework-roadmap](docs/design/agent-framework-roadmap.md)
- **Cost efficiency** — Mistral models are competitively priced vs GPT-4 and Claude

### Impact Areas

- [ ] API / Compatibility
- [ ] Security / Privacy
- [ ] Reliability / Availability
- [ ] Performance
- [ ] Documentation
- [ ] Unsure

### Files to Create
- [modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4s/modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala:0:0-0:0) — Provider client using Java `HttpClient`, OpenAI-compatible request/response format
- [modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralClientSpec.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4s/modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralClientSpec.scala:0:0-0:0) — Local HTTP server tests (success, error codes, edge cases)
- [modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralClientClosedStateTest.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4s/modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralClientClosedStateTest.scala:0:0-0:0) — Closed state handling tests
### Files to Modify
- [modules/core/src/main/scala/org/llm4s/llmconnect/config/ProviderConfig.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4s/modules/core/src/main/scala/org/llm4s/llmconnect/config/ProviderConfig.scala:0:0-0:0) — Add `MistralConfig` case class + companion with `DEFAULT_BASE_URL` and model-specific context windows
- [modules/core/src/main/scala/org/llm4s/llmconnect/provider/LLMProvider.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4s/modules/core/src/main/scala/org/llm4s/llmconnect/provider/LLMProvider.scala:0:0-0:0) — Add `case object Mistral`, update `name`, `all`, `fromName`
- [modules/core/src/main/scala/org/llm4s/llmconnect/LLMConnect.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4s/modules/core/src/main/scala/org/llm4s/llmconnect/LLMConnect.scala:0:0-0:0) — Register `MistralConfig → MistralClient` in both `buildClient` and `getClient`
- [modules/core/src/test/scala/org/llm4s/llmconnect/provider/LLMProviderSpec.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4s/modules/core/src/test/scala/org/llm4s/llmconnect/provider/LLMProviderSpec.scala:0:0-0:0) — Update provider count and add Mistral tests
- [modules/core/src/test/scala/org/llm4s/llmconnect/LLMConnectProviderTypeSafetyTest.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4s/modules/core/src/test/scala/org/llm4s/llmconnect/LLMConnectProviderTypeSafetyTest.scala:0:0-0:0) — Add Mistral type safety test
- [modules/samples/src/main/scala/org/llm4s/samples/metrics/PrometheusMetricsExample.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4s/modules/samples/src/main/scala/org/llm4s/samples/metrics/PrometheusMetricsExample.scala:0:0-0:0) — Add `MistralConfig` case to exhaustive match

### Impact Scope

Agent Framework

### Breaking Changes

No - Backward compatible

### Related Issues or PRs

#662 

### Implementation Willingness

- [x] I would like to implement this enhancement myself
- [ ] I would be open to collaborating on implementation
- [ ] I would prefer someone else to implement this

### Pre-submission Checklist

- [x] I have searched existing issues and found no duplicates
- [x] I have clearly described the current behavior
- [x] I have explained the proposed improvement
- [x] I have identified the benefits and impact

### PR DESCRIPTION
**Pull request by mandeepsingh2007**
## What does this PR do?

Adds **Mistral AI** as a new LLM provider to llm4s (Phase 4.2 — Provider Expansion).

Mistral uses an OpenAI-compatible `/v1/chat/completions` API, so the client follows the same proven pattern as CohereClient and other existing providers.

### Changes:
- **`MistralConfig`** — new config case class with model-specific context window fallbacks (supports mistral-large, mistral-small, codestral, open-mistral)
- **`MistralClient`** — new provider client extending `LLMClient` with `MetricsRecording`, handling non-streaming chat completions
- **`LLMProvider.Mistral`** — new case object registered in the sealed trait, `fromName`, and `all` list
- **`LLMConnect`** — wired MistralClient into both `buildClient` and `getClient` factory methods
- **`PrometheusMetricsExample`** — added `MistralConfig` to exhaustive match
- **Tests** — `MistralClientSpec` (7 tests), `MistralClientClosedStateTest` (4 tests), updated `LLMConnectProviderTypeSafetyTest` and `LLMProviderSpec`

## Related issue
<-- Link the issue this addresses: Fixes #660 

Relates to Phase 4.2 (Provider Expansion) in `agent-framework-roadmap.md` — adds Mistral alongside existing Cohere and Gemini providers.

## How was this tested?

- `sbt scalafmtAll` — all code formatted
- `sbt +test` — **all tests pass** on both Scala 2.13 and 3.x
- New tests added:
  - `MistralClientSpec` — successful response parsing, whitespace handling, missing text validation, HTTP 401→AuthenticationError, 429→RateLimitError, 500→ServiceError, nested error format
  - `MistralClientClosedStateTest` — closed state error, streaming not supported, idempotent close, model name in error message
  - `LLMConnectProviderTypeSafetyTest` — Mistral provider returns MistralClient
  - `LLMProviderSpec` — name, fromName, case-insensitive parsing, pattern matching for Mistral

## Checklist

- [x] I have read the [Contributing Guide](https://llm4s.org/reference/contributing)
- [x] PR is small and focused — one change, one reason
- [x] `sbt scalafmtAll` — code is formatted
- [x] `sbt +test` — tests pass on both Scala 2.13 and 3.x
- [x] New code includes tests
- [x] No unrelated changes included (branched from `main`, not from another PR)
- [x] Commit messages explain the "why"

### CODE REVIEW

**Comment by mandeepsingh2007** [comment]
@rorygraves @vim89 please review this pr

**Comment by arnavsharma990** [comment]
hey @mandeepsingh2007 , thanks for adding Mistral support and the accompanying tests.

**Change required:** the PR is currently failing the Codecov patch check (≈69% vs 80% required), mainly due to a few uncovered new branches in `MistralConfig` (model context-window fallback cases) and some remaining fallback/error paths.

Adding a small `MistralConfigSpec` (to exercise the fallback model mappings + validation requires) and one or two extra edge case tests should bring patch coverage back above the threshold, then it would be good to be reviewed

**Comment by mandeepsingh2007** [comment]
@rorygraves @vim89 please review this pr

**Comment by arnavsharma990** [comment]
Awesome , thanks for addressing the coverage gaps so quickly! Patch coverage looks great now (~95%) 
Ready for further review from the mentors/maintainers.

File: modules/core/src/main/scala/org/llm4s/llmconnect/config/ProviderConfig.scala [hunk 1]
```diff
@@ -503,3 +503,68 @@ object CohereConfig {
...
+
+    registryResult match {
+      case Some(metadata) =>
+        val contextWindow = metadata.maxInputTokens.getOrElse(DefaultContextWindow)
+        val reserve       = metadata.maxOutputTokens.getOrElse(DefaultReserveCompletion)
+        logger.debug(s"Using ModelRegistry metadata for $modelName: context=$contextWindow, reserve=$reserve")
+        (contextWindow, reserve)
+      case None =>
+        logger.debug(s"Model $modelName not found in registry, using fallback values")
+        modelName.toLowerCase match {
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/config/ProviderConfig.scala]: Use locale-stable lowercasing here (`toLowerCase(Locale.ROOT)`) to avoid locale-dependent fallback matching behavior. Principle: deterministic configuration parsing must not depend on host locale.

File: modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+      case AssistantMessage(contentOpt, _) =>
+        contentOpt.filter(_.nonEmpty).map { c =>
+          ujson.Obj(
+            "role"    -> "assistant",
+            "content" -> c
+          )
+        }
+
+      case _ =>
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala]: When tool-calling is unsupported, silently dropping non-user/system messages here can produce incorrect prompts and hard-to-debug behavior. Please fail fast with a clear validation/config error instead of omitting message context. Principle: unsupported capability should be explicit, not silently degraded. This is a low-level + high-level design integrity concern.

File: modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+      json.obj
+        .get("message")
+        .flatMap(_.strOpt)
+        .orElse(
+          json.obj
+            .get("error")
+            .flatMap(v => v.obj.get("message").flatMap(_.strOpt).orElse(v.strOpt))
+        )
+        .getOrElse(body)
+    }.getOrElse(body)
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala]: This fallback surfaces raw provider body in the returned error path. Please sanitize/truncate before exposing it to callers to avoid accidental leakage of upstream internals. Principle: error handling should preserve context without leaking sensitive payloads.

File: modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralClientSpec.scala [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+        |        "content": "Response after skipped message"
+        |      }
+        |    }
+        |  ],
+        |  "usage": {
+        |    "prompt_tokens": 10,
+        |    "completion_tokens": 5,
+        |    "total_tokens": 15
+        |  }
+        |}""".stripMargin
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralClientSpec.scala:567]: There are many `toOption.get` and `left.toOption.get` usages across this spec. Please switch to `fold`/pattern-safe assertions so the tests stay total and avoid partial access. Principle: avoid unsafe partial functions even in tests.

**Comment by mandeepsingh2007** [comment]
@vim89 @rorygraves  please review it i have fixed all changes

File: modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralClientClosedStateTest.scala [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+  "MistralClient" should "return ConfigurationError when complete() is called after close()" in {
+    val client = new MistralClient(createTestConfig)
+
+    client.close()
+
+    val result = client.complete(createTestConversation, CompletionOptions())
+
+    result.isLeft shouldBe true
+    result.left.toOption.get shouldBe a[ConfigurationError]
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralClientClosedStateTest.scala]: This still uses unsafe partial assertions with left.toOption.get. Previous review asked to make assertions total. Please switch this spec to fold or pattern-safe matching end to end. Principle: tests must be deterministic and free from partial access.

File: modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+      case other =>
+        Left(
+          ValidationError(
+            "conversation",
+            s"Mistral does not support message type: ${other.getClass.getSimpleName}"
+          )
+        )
+    }
+
+    results.collectFirst { case Left(err) => Left(err) }.getOrElse {
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala]: This uses getOrElse in core flow and introduces a null-marker pattern in the same function. Please keep this in pure transformations using map flatMap fold and avoid getOrElse based fallback branching. Principle: prefer total monadic transformations over fallback extraction patterns.

File: modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+              ujson.Obj(
+                "role"    -> "assistant",
+                "content" -> c
+              )
+            )
+          case None =>
+            Right(ujson.Null) // marker for empty assistant messages
+        }
+
+      case other =>
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala:158]: The fail-fast branch for unsupported message types is a good fix, but it needs an explicit test case in this PR so the behavior stays protected. Please add a spec using an unsupported message type and assert the ValidationError text. Principle: capability boundaries must be tested, not only implemented.

**Review by vim89 — CHANGES_REQUESTED** [review]
Request changes: provider expansion value is good and this PR is now much closer to merge after fixing locale-stable matching, explicit unsupported message handling, and safer upstream error detail handling.

Roadmap and issue alignment are strong for provider parity and reliability, and the core wiring across config, provider enum, and client factory is coherent.
Remaining blocker is verification quality: closed-state spec still has partial assertions, and the new unsupported-message fail-fast path is not explicitly covered by a dedicated test.
There is also one remaining idiomatic gap in core flow where getOrElse plus null-marker fallback is used instead of total monadic transformations. 

Please close these and this should be in merge-ready shape.

I did not run tests or build.

**Comment by mandeepsingh2007** [comment]
@vim89 @rorygraves please re-review this pr

File: modules/core/src/main/scala/org/llm4s/llmconnect/config/ProviderConfig.scala [hunk 8]
```diff
@@ -503,3 +505,68 @@ object CohereConfig {
...
+        val reserve       = metadata.maxOutputTokens.getOrElse(DefaultReserveCompletion)
+        logger.debug(s"Using ModelRegistry metadata for $modelName: context=$contextWindow, reserve=$reserve")
+        (contextWindow, reserve)
+      case None =>
+        logger.debug(s"Model $modelName not found in registry, using fallback values")
+        modelName.toLowerCase(Locale.ROOT) match {
+          case name if name.contains("mistral-large")  => (128000, DefaultReserveCompletion)
+          case name if name.contains("mistral-medium") => (32000, DefaultReserveCompletion)
+          case name if name.contains("mistral-small")  => (128000, DefaultReserveCompletion)
+          case name if name.contains("codestral")      => (256000, DefaultReserveCompletion)
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/config/ProviderConfig.scala]: This fallback is likely stale against current Mistral model docs where recent Codestral variants are 128k context. Overestimating context can trigger avoidable upstream validation failures. Principle: provider limits must be source-of-truth accurate and conservative by default.

File: modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralConfigSpec.scala [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  it should "return 128000 for mistral-small models (fallback)" in {
+    val cfg =
+      MistralConfig.fromValues("custom-mistral-small-v1", apiKey = "key", baseUrl = MistralConfig.DEFAULT_BASE_URL)
+    cfg.contextWindow shouldBe 128000
+  }
+
+  it should "return 256000 for codestral models (fallback)" in {
+    val cfg = MistralConfig.fromValues("custom-codestral-v1", apiKey = "key", baseUrl = MistralConfig.DEFAULT_BASE_URL)
+    cfg.contextWindow shouldBe 256000
+  }
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralConfigSpec.scala:60]: This expectation bakes in the same risky 256k fallback. Please align with the updated conservative limit or registry-backed behavior so config tests do not lock incorrect provider constraints. Principle: tests must protect real provider contracts, not outdated assumptions.

File: modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralConfigSpec.scala [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+  it should "return 128000 for mistral-small models (fallback)" in {
+    val cfg =
+      MistralConfig.fromValues("custom-mistral-small-v1", apiKey = "key", baseUrl = MistralConfig.DEFAULT_BASE_URL)
+    cfg.contextWindow shouldBe 128000
+  }
+
+  it should "return 256000 for codestral models (fallback)" in {
+    val cfg = MistralConfig.fromValues("custom-codestral-v1", apiKey = "key", baseUrl = MistralConfig.DEFAULT_BASE_URL)
+    cfg.contextWindow shouldBe 256000
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralConfigSpec.scala]: Blocking: this test hard-codes `256000` for generic codestral fallback and cements the mismatch above. Please validate against the canonical model-limit source so tests stay correct when provider metadata evolves. Principle: keep config fallbacks and verification anchored to one source of truth.

**Comment by mandeepsingh2007** [comment]
@vim89 @rorygraves please re-review this pr

File: modules/core/src/main/scala/org/llm4s/llmconnect/config/ProviderConfig.scala [hunk 11]
```diff
@@ -503,3 +505,68 @@ object CohereConfig {
...
+        val reserve       = metadata.maxOutputTokens.getOrElse(DefaultReserveCompletion)
+        logger.debug(s"Using ModelRegistry metadata for $modelName: context=$contextWindow, reserve=$reserve")
+        (contextWindow, reserve)
+      case None =>
+        logger.debug(s"Model $modelName not found in registry, using fallback values")
+        modelName.toLowerCase(Locale.ROOT) match {
+          case name if name.contains("mistral-large")  => (128000, DefaultReserveCompletion)
+          case name if name.contains("mistral-medium") => (32000, DefaultReserveCompletion)
+          case name if name.contains("mistral-small")  => (128000, DefaultReserveCompletion)
+          case name if name.contains("codestral")      => (128000, DefaultReserveCompletion)
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/config/ProviderConfig.scala]: Fallback values here are still detached from canonical in-repo model metadata; for `codestral` and `mistral-small` the metadata currently resolves much lower limits.
Please keep fallback conservative and anchored to a single source of truth to avoid context-window overestimation at runtime.
Principle: provider limits should be deterministic and metadata-driven, not keyword-hardcoded drift.

File: modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralConfigSpec.scala [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+  it should "return 128000 for mistral-small models (fallback)" in {
+    val cfg =
+      MistralConfig.fromValues("custom-mistral-small-v1", apiKey = "key", baseUrl = MistralConfig.DEFAULT_BASE_URL)
+    cfg.contextWindow shouldBe 128000
+  }
+
+  it should "return 128000 for codestral models (fallback)" in {
+    val cfg = MistralConfig.fromValues("custom-codestral-v1", apiKey = "key", baseUrl = MistralConfig.DEFAULT_BASE_URL)
+    cfg.contextWindow shouldBe 128000
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/llmconnect/provider/MistralConfigSpec.scala]: This spec still hard-codes the same fallback assumptions, so it can lock drift instead of protecting provider truth.
Please derive assertions from the canonical metadata path (or validate fallback conservativeness explicitly) rather than pinning magic context numbers.
Principle: tests should enforce one authoritative source for provider limits.

File: modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        .flatMap(_.headOption)
+
+      val textOpt = choiceOpt
+        .flatMap(_.obj.get("message"))
+        .flatMap(_.obj.get("content"))
+        .flatMap(_.strOpt)
+        .map(_.trim)
+        .filter(_.nonEmpty)
+
+      val text = textOpt.getOrElse("")
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala]: Core parse path still uses fallback extraction (`getOrElse`) before final branching.
Please keep this flow fully monadic (`map`/`flatMap`/`fold`) to avoid mixed extraction+sentinel patterns in provider core.
Principle: total transformations keep error semantics explicit and deterministic.

**Comment by mandeepsingh2007** [comment]
@vim89 @rorygraves please re review this pr

File: modules/core/src/main/scala/org/llm4s/llmconnect/config/ProviderConfig.scala [hunk 14]
```diff
@@ -503,3 +505,68 @@ object CohereConfig {
...
+    val registryResult =
+      ModelRegistry
+        .lookup("mistral", modelName)
+        .toOption
+        .orElse(ModelRegistry.lookup(modelName).toOption)
+
+    registryResult match {
+      case Some(metadata) =>
+        val contextWindow = metadata.maxInputTokens.getOrElse(DefaultContextWindow)
+        val reserve       = metadata.maxOutputTokens.getOrElse(DefaultReserveCompletion)
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/config/ProviderConfig.scala:657]: Fallback mapping still depends on keyword heuristics and can drift from provider reality over time. Prefer a single metadata-backed source with conservative fallback only when registry lookup fails.
Principle: provider limits must be deterministic and sourced from one authority.
(Recent Mistral docs show Codestral context at 128k, so drift risk is real.)

File: modules/core/src/main/scala/org/llm4s/llmconnect/LLMConnect.scala [hunk 15]
```diff
@@ -57,6 +59,7 @@ object LLMConnect {
       case (LLMProvider.Gemini, cfg: GeminiConfig)       => GeminiClient(cfg, metrics)
       case (LLMProvider.DeepSeek, cfg: DeepSeekConfig)   => DeepSeekClient(cfg, metrics)
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/LLMConnect.scala:123]: Good wiring here, but provider activation is still incomplete from the standard config boundary: `ProviderConfigLoader` has no `mistral` branch yet, so `LLM_MODEL=mistral/...` remains unreachable through `Llm4sConfig.provider()`.
Principle: provider registration must be end-to-end across loader plus factory paths.

File: modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+
+    statusCode match {
+      case 401 | 403     => Left(AuthenticationError("mistral", details))
+      case 429           => Left(RateLimitError("mistral"))
+      case 400           => Left(ValidationError("request", details))
+      case s if s >= 500 => Left(ServiceError(s, "mistral", details))
+      case s             => Left(ServiceError(s, "mistral", details))
+    }
+  }
+
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/llmconnect/provider/MistralClient.scala:264]: This status matching is duplicated: `case s if s >= 500` and `case s` currently return the same `ServiceError`.
Principle: keep error routing explicit and minimal to avoid dead branches and drift.
Please collapse these branches.
